frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -1001,7 +1002,25 @@ func dispatchCopy(d *dispatchState, cfg copyConfig) error {
 
 	for _, src := range cfg.params.SourcePaths {
 		commitMessage.WriteString(" " + src)
-		if strings.HasPrefix(src, "http://") || strings.HasPrefix(src, "https://") {
+		if gitutil.IsGitRef(src) {
+			if !cfg.isAddCommand {
+				return errors.New("source can't be a git ref for COPY")
+			}
+			g, err := gitutil.ParseGitRef(src)
+			if err != nil {
+				return errors.Wrapf(err, "failed to parse a git ref %q", src)
+			}
+			st := llb.Git(g.Repo, g.Commit, llb.KeepGitDir())
+			opts := append([]llb.CopyOption{&llb.CopyInfo{
+				Mode:           mode,
+				CreateDestPath: true,
+			}}, copyOpt...)
+			if a == nil {
+				a = llb.Copy(st, "/", dest, opts...)
```

Comment:
We should make sure that

```
FROM scratch
ADD git:// / (or .)
```

Is converted to direct `llb.Git()`, not `llb.Scratch().File(llb.Copy(llb.Git(), "/", "/"))` . Possibly this could be also optimized in the llb libarary.

Reply:
Do we need to work on this in this PR or a separate PR?

Reply:
Would be better in this PR. Separate commit is ok. If you want to do this in `llb` client level then it can be done in parallel on separate PR but we should avoid the possibility of having the git feature but not this optimization merged.

Reply:
Do we have that optimization for `http` sources?

Reply:
Added an optimizer in the second commit